mlserver/parallel/pool.py

```diff
@@ -0,0 +1,207 @@
+            worker = Worker(self._requests, self._responses, model_updates)
+            worker.start()
+            self._workers[worker.pid] = worker
+
+        # Start processing responses
+        self._start_processing_responses()
+
+    def _start_processing_responses(self):
+        self._active = True
+        self._process_responses_task = asyncio.create_task(self._process_responses())
+        self._process_responses_task.add_done_callback(self._process_responses_cb)
+
+    def _process_responses_cb(self, process_responses):
+        try:
+            process_responses.result()
+        except Exception:
+            logger.exception("Response processing loop crashed. Restarting the loop...")
+            # If process loop crashed, restart it
+            self._start_processing_responses()
+
+    async def _process_responses(self):
+        logger.debug("Starting response processing loop...")
+        loop = asyncio.get_event_loop()
+        while self._active:
+            response = await loop.run_in_executor(None, self._responses.get)
+            # If the queue gets terminated, detect the "sentinel value" and
+            # stop reading
+            if response is END_OF_QUEUE:
+                return
+
+            await self._process_response(response)
+
+    async def _process_response(self, response: InferenceResponseMessage):
+        internal_id = response.id
+
+        async_response = self._async_responses[internal_id]
+
+        if response.inference_response:
+            async_response.set_result(response.inference_response)
+        elif response.exception:
+            async_response.set_exception(response.exception)
+        else:
+            exc = InferenceError("Inference returned no value")
+            async_response.set_exception(exc)
+
+    async def predict(
+        self, model_settings: ModelSettings, inference_request: InferenceRequest
+    ) -> InferenceResponse:
+        internal_id = generate_uuid()
+
+        model_version = None
+        if model_settings.parameters:
+            model_version = model_settings.parameters.version
+
+        request_message = InferenceRequestMessage(
+            id=internal_id,
+            model_name=model_settings.name,
+            model_version=model_version,
+            inference_request=inference_request,
+        )
+        self._requests.put(request_message)
+
+        loop = asyncio.get_running_loop()
+        async_response = loop.create_future()
+        self._async_responses[internal_id] = async_response
+
+        return await self._wait_response(internal_id)
+
+    async def _wait_response(self, internal_id: str) -> InferenceResponse:
+        async_response = self._async_responses[internal_id]
+
+        try:
+            inference_response = await async_response
+            return inference_response
+        finally:
+            del self._async_responses[internal_id]
+
+        return await async_response
```

Comment:
Not necessarily. The `finally` block should get called every time the `try` block finalises (i.e. either because of a succesful run or because of an exception), so it just ensures that the async response always gets removed from the dict of pending ones.